### ISSUE TITLE
systemd-sysusers fails to edit /etc/group without /proc (under dnf --installroot)

### ISSUE THREAD

**Issue by mikhailnov** [systemd/systemd#18276]
I am running the script `mkimage-dnf.sh` which builds a rootfs with systemd
https://github.com/mikhailnov/docker-rosa/blob/7d5d0951d648a08bdaa9aed7d3e41bd6310c2cb9/mkimage-dnf.sh
(this link is to the exact revision that I am using when writing this)
`sudo arch=x86_64 addPackages=bind rootfsPackTarball=0 rootfsPackSquash=1 ./mkimage-dnf.sh`
(I may also add `privateUsersOffset=150000000` to further use this image with `systemd-nspawn --private-users=150000000`)

This script runs `dnf --installroot=<empty directory> install basesystem-minimal <...>`.

Some packages, e.g. `bind`, create users with `systemd-sysusers` inside `%pre`/`%post` scriptlets.

`systemd-sysusers` fails to create group inside the dnf transaction:
```
  Запуск скриптлета: bind-1:9.11.18-34.x86_64                                                                                                                                         237/248 
Creating group kvm with gid 494.
Creating group render with gid 493.
Creating group ldap with gid 492.
Creating group named with gid 491.
Creating user ldap (OpenLDAP server (ldap) user) with uid 492 and gid 492.
Creating user named (BIND DNS Server) with uid 491 and gid 491.
Failed to write files: Function not implemented

  Установка        : bind-1:9.11.18-34.x86_64                                                                                                                                         237/248 
предупреждение: group named does not exist - using root
предупреждение: group named does not exist - using root
```

Now I chroot into it and try to run `systemd-sysusers` manually:

```
[user@rosa2019 docker-rosa]$ sudo chroot BUILD_rootfs-std-rosa2019.1_x86_64_2021-01-16
[sudo] пароль для user: 
[root@rosa2019 /]# export LC_ALL=C
[root@rosa2019 /]# rpm -ql bind | grep sysusers
/usr/lib/sysusers.d/named.conf
[root@rosa2019 /]# cat /usr/lib/sysusers.d/named.conf
u named - "BIND DNS Server" /var/lib/named
g named - -
m named named
[root@rosa2019 /]# systemd-sysusers named.conf
Creating group named with gid 494.
Creating user named (BIND DNS Server) with uid 494 and gid 494.
Failed to write files: Function not implemented
[root@rosa2019 /]# 
```

The error has been reproduced.

Let's have a look at current `/etc/passwd` and `/etc/group`:

```
[root@rosa2019 /]# tail /etc/passwd
rpm:x:16:16:rpm:/var/lib/rpm:/bin/false
messagebus:x:28:28:nscd:/:/sbin/nologin
nscd:x:29:29:nscd:/:/sbin/nologin
ftp:x:14:50:FTP User:/var/ftp:/sbin/nologin
nobody:x:65534:65534:Kernel Overflow User:/:/sbin/nologin
systemd-timesync:x:499:499:systemd Time Synchronization:/:/sbin/nologin
systemd-network:x:498:498:systemd Network Management:/:/sbin/nologin
systemd-resolve:x:497:497:systemd Resolver:/:/sbin/nologin
systemd-bus-proxy:x:496:496:systemd Bus Proxy:/:/sbin/nologin
systemd-coredump:x:495:495:systemd Core Dumper:/var/lib/systemd/coredump:/sbin/nologin
[root@rosa2019 /]# tail /etc/group 
input:x:101:
nogroup:x:65533:
nobody:x:65534:
nopasswd:x:999:
systemd-journal:x:190:
systemd-timesync:x:499:
systemd-network:x:498:
systemd-resolve:x:497:
systemd-bus-proxy:x:496:
systemd-coredump:x:495:
[root@rosa2019 /]# 
```

`systemd-*` stuff has been created by `useradd` and `groupadd` inside scriptlets of `systemd` RPM package, not by `systemd-sysusers`.

Now I edit `/etc/yum.repos.d/rosa-x86_64.repo` and enable `rosa-x86_64-debuginfo` repository and run `dnf install systemd-debuginfo gdb strace` inside the chroot.

systemd version is v246.9, commit https://github.com/systemd/systemd/commit/62c03398ba6012ff693fbce24c142309cdbc1cd0 ("sysusers: modernize file backup logic a bit") has been cherry-picked, but it has not resolved the issue.

Let's run the same inside the chroot under `strace`:

```
[root@rosa2019 /]# strace systemd-sysusers named.conf
<...>
openat(AT_FDCWD, "/etc/group", O_RDONLY|O_CLOEXEC) = 4
fstat(4, {st_mode=S_IFREG|0644, st_size=627, ...}) = 0
lseek(4, 0, SEEK_SET)                   = 0
read(4, "root:x:0:\nbin:x:1:\ndaemon:x:2:\ns"..., 4096) = 627
read(4, "", 4096)                       = 0
close(4)                                = 0
writev(2, [{iov_base="Creating user named (BIND DNS Se"..., iov_len=63}, {iov_base="\n", iov_len=1}], 2Creating user named (BIND DNS Server) with uid 494 and gid 494.
) = 64
umask(077)                              = 022
getpid()                                = 148564
openat(AT_FDCWD, "/etc/.#groupax8dsU", O_RDWR|O_CREAT|O_EXCL|O_CLOEXEC, 0600) = 4
umask(022)                              = 077
fcntl(4, F_GETFL)                       = 0x8002 (flags O_RDWR|O_LARGEFILE)
openat(AT_FDCWD, "/etc/group", O_RDONLY|O_CLOEXEC) = 5
fstat(5, {st_mode=S_IFREG|0644, st_size=627, ...}) = 0
fstat(4, {st_mode=S_IFREG|0600, st_size=0, ...}) = 0
chmod("/proc/self/fd/4", 0644)          = -1 ENOENT (No such file or directory)
statfs("/proc/", {f_type=BTRFS_SUPER_MAGIC, f_bsize=4096, f_blocks=47909632, f_bfree=5774229, f_bavail=5336481, f_files=0, f_ffree=0, f_fsid={val=[3967447838, 2986924096]}, f_namelen=255, f_frsize=4096, f_flags=ST_VALID|ST_RELATIME}) = 0
unlink("/etc/.#groupax8dsU")            = 0
close(4)                                = 0
close(5)                                = 0
writev(2, [{iov_base="Failed to write files: Function "..., iov_len=47}, {iov_base="\n", iov_len=1}], 2Failed to write files: Function not implemented
) = 48
close(3)                                = 0
exit_group(1)                           = ?
+++ exited with 1 +++
```

We see above that `/proc` is BTRFS, it is because `/proc` is not mounted in the chroot (note that I just ran `chroot <directory>` after `dnf` had already finished running). Let's mount it:
`mount -t proc proc /proc`

And now `systemd-sysusers` works correctly:

```
[root@rosa2019 /]# systemd-sysusers named.conf
Creating group named with gid 494.
Creating user named (BIND DNS Server) with uid 494 and gid 494.
[root@rosa2019 /]#
```

`gdb` seems to not work properly without `/proc`...

**Comment by mikhailnov** [comment]
Commit https://github.com/mikhailnov/docker-rosa/commit/600f8fa36f169b9e476cf4e67b463ab5f6c9955f which added mounting `/proc` before running `dnf` has worked around this problem, but I think that systemd-sysusers must work without `/proc`.

**Comment by poettering** [comment]
Sorry, but no. /proc is half of the Linux program API, we can#t work without that. Please fix your environment to just mount that. What's stopping you?

This smells like you are creating yoruself problems just for the sake of creating problems.

I am tempted to just close this tbh.

**Comment by mikhailnov** [comment]
Hm, ok, there is no problem to mount /proc, let's close it then, thanks. Why I opened this - /proc does not exist in a pure chroot environment.

**Comment by poettering** [comment]
btw, we previously added some patches to make this more discoverable (see f8606626ed3c2582e06543550d58fe9886cdca5f for example), we can always build on that. But I am not convinced we should work on making our toolset work without /proc/ at all

**Comment by mikhailnov** [comment]
The problem is that, first of all, it would be nice to make chroot buildable without /proc being mounted separately, secondly, the error message "Failed to write files: Function not implemented" is not clear and does not tell that /proc is needed.

We in ROSA Linux (RPM) are migrating most of the packages to config-based sysusers.d instead of creating users and groups by scripts in %pre, because in the second case it is not clear why a user or group was created. I would like to avoid requiring /proc to boostrap a chroot, but it is not a critical issue to my mind.

Failure of tmpfiles when bootstrapping a chroot is not critical because files will be created when system is booted (well, if it a container image which will never be booted, then may be a problem), but in case of sysusers a user+group must be created before RPM package which has files owned by user:group (%attr) is installed.

**Comment by boucman** [comment]
isn't that a symptom of a deeper bug ? 
If I am cross-building a system and invoking systems-sysuser with --image or --root, the image is probably or a totally different system, and whatever info is extracted from /proc is probably wrong in the first place. no ?

**Comment by poettering** [comment]
@boucman we use `/proc/self/fd/<fd>` all the time as a way to reopen fds, or to dereference an O_PATH fd and so on. That has nothing to do with --image= or root=, it's just how you work with fds on Linux.

**Comment by mikhailnov** [comment]
19.01.2021 20:29, Boucman пишет:
>
> isn't that a symptom of a deeper bug ?
> If I am cross-building a system and invoking systems-sysuser with --image or --root, the image is probably or a totally different system, and whatever info is extracted from /proc is probably wrong in the first place. no ?
>
In this case systemd-sysusers is being run on your system, /proc from it is being used, because it does not chroot into the target system, as far as I understand.

**Comment by boucman** [comment]
ok, thx

**Comment by socketpair** [comment]
Resolved in https://github.com/systemd/systemd/commit/82855aa610510ddacf72bec03629839a7829e3b7

**Comment by mikhailnov** [comment]
> @mikhailnov
> Resolved in [82855aa](https://github.com/systemd/systemd/commit/82855aa610510ddacf72bec03629839a7829e3b7)

thanks for noting!
